File: examples/e2e-consent-demo/sp-shared/audit.ts
// Service-Provider-side audit emission.
//
// The SP is the only party that knows several facts the activity trail needs:
// whether a presentation actually verified, why it did not, and whether the
// scope a tool requires was present in the grant. Recording those from the
// agent would mean trusting the agent's account of its own authorization —
// so the SP reports them itself, from the process that made the decision.
//
// Every emission is best-effort. An SP must keep serving tool calls when the
// audit sink is unreachable: a booking is not less valid because a log write
// failed, and the failure must never surface to the caller.

import type { AuditEventType } from '@helixid/core';

/**
 * The shared activity-trail envelope. Field names are fixed even though almost
 * everything is optional — that is what lets issuance, presentation,
 * verification, authorization and invocation be queried as one trail later.
 */
export interface ActivityEvent {
  event: AuditEventType;
  timestamp?: string;
  correlationId?: string | undefined;
  agentDid?: string | undefined;
  userDid?: string | undefined;
  vcId?: string | undefined;
  credentialType?: string | undefined;
  issuer?: string | undefined;
  scopes?: string[] | undefined;
  validUntil?: string | undefined;
  credentialStatus?: string | undefined;
  serviceDid?: string | undefined;
  serviceName?: string | undefined;
  toolName?: string | undefined;
  requiredScope?: string | undefined;
  effectiveScopes?: string[] | undefined;
  vpId?: string | undefined;
  result?: 'success' | 'failure' | 'blocked' | undefined;
  reason?: string | undefined;
  resultSummary?: string | undefined;
}

export interface AuditEmitter {
  emit(event: ActivityEvent): void;
}

export interface AuditEmitterOptions {
  helixApiUrl?: string | undefined;
  adminApiKey?: string | undefined;
  serviceDid: string;
  serviceName: string;
  /** Called with a one-line description of anything that could not be sent. */
  onError?: (message: string) => void;
}

/** No-op emitter, used when the SP has no audit sink configured. */
const NOOP: AuditEmitter = { emit: () => undefined };

export function createAuditEmitter(options: AuditEmitterOptions): AuditEmitter {
  const { helixApiUrl, adminApiKey, serviceDid, serviceName } = options;
  if (!helixApiUrl || !adminApiKey) return NOOP;

  const endpoint = `${helixApiUrl.replace(/\/$/, '')}/v1/audit-log/events`;

  return {
    emit(event: ActivityEvent): void {
      // Deliberately not awaited: the tool call's latency and its success must
      // not depend on the audit sink. Errors are swallowed after logging.
      void fetch(endpoint, {
        method: 'POST',
        headers: { 'content-type': 'application/json', 'x-admin-api-key': adminApiKey },
        body: JSON.stringify({
          serviceDid,
          serviceName,
          source: 'sp',
          timestamp: new Date().toISOString(),
          ...event,
        }),
      })
        .then(async (response) => {
          if (!response.ok) {
            options.onError?.(`audit ${event.event} rejected: HTTP ${response.status}`);
          }
        })
        .catch((error: unknown) => {
          options.onError?.(`audit ${event.event} failed: ${(error as Error).message}`);
        });
    },
  };
}

File: examples/e2e-consent-demo-py/sp_shared/audit.py
# Service-Provider-side audit emission. Python port of sp-shared/audit.ts.
#
# Every emission is best-effort and fire-and-forget (a background thread, not
# awaited) -- an SP must keep serving tool calls when the audit sink is
# unreachable.

from __future__ import annotations

import threading
from datetime import datetime, timezone
from typing import Any, Callable, Optional

import requests

# The demo's audit event type strings, mirroring @helixid/sdk-js's AuditEvents.
class AuditEvents:
    VC_ISSUED = "VC_ISSUED"
    VC_PRESENTED = "VC_PRESENTED"
    VP_VERIFIED = "VP_VERIFIED"
    VP_REJECTED = "VP_REJECTED"
    AUTHZ_GRANTED = "AUTHZ_GRANTED"
    AUTHZ_DENIED = "AUTHZ_DENIED"
    TOOL_INVOKED = "TOOL_INVOKED"


class AuditEmitter:
    def emit(self, event: dict) -> None:  # pragma: no cover - overridden
        raise NotImplementedError


class _NoopEmitter(AuditEmitter):
    def emit(self, event: dict) -> None:
        return None


NOOP = _NoopEmitter()


class _HttpAuditEmitter(AuditEmitter):
    def __init__(
        self,
        endpoint: str,
        admin_api_key: str,
        service_did: str,
        service_name: str,
        on_error: Optional[Callable[[str], None]] = None,
    ) -> None:
        self._endpoint = endpoint
        self._admin_api_key = admin_api_key
        self._service_did = service_did
        self._service_name = service_name
        self._on_error = on_error

    def emit(self, event: dict) -> None:
        payload = {
            "serviceDid": self._service_did,
            "serviceName": self._service_name,
            "source": "sp",
            "timestamp": datetime.now(timezone.utc).isoformat().replace("+00:00", "Z"),
            **event,
        }

        def _send() -> None:
            try:
                res = requests.post(
                    self._endpoint,
                    json=payload,
                    headers={"x-admin-api-key": self._admin_api_key},
                    timeout=5,
                )
                if not res.ok and self._on_error:
                    self._on_error(f"audit {event.get('event')} rejected: HTTP {res.status_code}")
            except Exception as exc:  # noqa: BLE001 - best-effort sink
                if self._on_error:
                    self._on_error(f"audit {event.get('event')} failed: {exc}")

        threading.Thread(target=_send, daemon=True).start()


def create_audit_emitter(
    helix_api_url: Optional[str],
    admin_api_key: Optional[str],
    service_did: str,
    service_name: str,
    on_error: Optional[Callable[[str], None]] = None,
) -> AuditEmitter:
    if not helix_api_url or not admin_api_key:
        return NOOP
    endpoint = f"{helix_api_url.rstrip('/')}/v1/audit-log/events"
    return _HttpAuditEmitter(endpoint, admin_api_key, service_did, service_name, on_error)
